logger/zerolog_logger.go

```diff
@@ -0,0 +1,105 @@
+package logger
+
+import (
+	"os"
+
+	"github.com/rs/zerolog"
+)
+
+// NewZeroLogger returns a new ZeroLogger
+func NewZeroLogger() *ZeroLogger {
+	l := zerolog.New(os.Stderr).With().Timestamp().Logger()
+	l = l.Output(zerolog.ConsoleWriter{Out: os.Stderr})
+	return &ZeroLogger{Logger: l}
+}
+
+// Compile time assertion that our logger meets the intferface specifications
+var _ Generic = (*ZeroLogger)(nil)
+
+// ZeroLogger is a wrapper around zerolog so it can implement our interface
+type ZeroLogger struct {
+	Logger zerolog.Logger
+}
+
+// Debug calls the debug method of the registered logger
+func (zl *ZeroLogger) Debug(msg string, keyvals ...interface{}) {
+	if !validateKeyvals(keyvals...) {
+		zl.Logger.Error().Str("debug_message", msg).
+			Msg("Debug logger function keyvals are not valid, may be missing context...")
+		return
+	}
+
+	addFields(zl.Logger.Debug(), keyvals...).Msg(msg)
+}
+
+// Info calls the info method of the registered logger
+func (zl *ZeroLogger) Info(msg string, keyvals ...interface{}) {
+	if !validateKeyvals(keyvals...) {
+		zl.Logger.Error().Str("info_message", msg).
+			Msg("Info logger function keyvals are not valid, may be missing context...")
+		return
+	}
+
+	addFields(zl.Logger.Info(), keyvals...).Msg(msg)
+}
+
+// Warn calls the info method of the registered logger
+func (zl *ZeroLogger) Warn(msg string, keyvals ...interface{}) {
+	if !validateKeyvals(keyvals...) {
+		zl.Logger.Error().Str("warn_message", msg).
+			Msg("Warn logger function keyvals are not valid, may be missing context...")
+		return
+	}
+
+	addFields(zl.Logger.Warn(), keyvals...).Msg(msg)
+}
+
+// Error calls the error method of the registered logger
+func (zl *ZeroLogger) Error(msg string, keyvals ...interface{}) {
+	if !validateKeyvals(keyvals...) {
+		zl.Logger.Error().Str("error_message", msg).
+			Msg("Error logger function keyvals are not valid, may be missing context...")
+		return
+	}
+
+	addFields(zl.Logger.Error(), keyvals...).Msg(msg)
+}
+
+// With calls the with method of the registered logger, and returns
+// a logger with those fields attached by default
+func (zl *ZeroLogger) With(keyvals ...interface{}) Generic {
+	if !validateKeyvals(keyvals...) {
+		zl.Logger.Error().
+			Msg("Provided keyvals don't match correct pattern for With() method")
+		return zl
+	}
+
+	l := zl.Logger.With()
+	for i := 0; i < len(keyvals); i += 2 {
```

Comment:
I'm not really sure what `keyvals` is meant to represent. Anyways it seems like the purpose is to allow arbitrary pairs of `string:<something else>`. Why not just make a struct type for that instead of relying on this `+= 2` stuff? That also makes me have to read the source of these methods to understand what this `keyvals` argument is even used for. Then just take an arbitrary number of those structs as this argument instead.

Reply:
The point of it is to allow easier logging so that rather than doing 

```go
log.Info("my message", NewLogFields("key", val))
```

We have an easy to write
```go
log.Info("msg", "err", error")
```

Plus the only benefit of using a struct besides clarity (which I think can be fixed by better documentation) would be a compile time assertion that the call to the log method had the correct key value pairs, and I can't think of a nice way to do that in one or two lines while still keeping the generic nature of the interface.

Reply:
I think the other thing to keep in mind is this logger interface is meant to stay internal so these methods should only be used by internal code, the only reason it is exposed is so the user can *replace* it if they want to, which likely will require a read of the source code anyway. By doing it this way it makes our life easier when adding log statements to legion. 

If you can think of a nice way to implement a struct in a way that keeps a log statement in a single line, and can give us compile time assertions I'd be happy to implement it.

Reply:
Ok after some thinking I updated the Logger interface to be a chaining one. Now the logger interface works like:
 ```go
logger.Debug().Field(key string, val interface{}).Log(msg string)
```
So now we have a compile time check that the fields have matching key value pairs (so no runtime failures anymore)

Reply:
I like that a lot better 👍